src/lib/RTC.ts

```diff
@@ -0,0 +1,249 @@
+    customReceiveMessageHandler: any;
+    attemptReconnectInterval: any;
+
+    constructor() {
+        this.localConnection = new RTCPeerConnection();
+        this.sendChannel = null;
+        this.receiveChannel = null;
+        this.customSendChannelStatusHandler = null;
+        this.customReceiveChannelStatusHandler = null;
+        this.customReceiveMessageHandler = null;
+        this.attemptReconnectInterval = null;
+
+        socket.on(Constants.RTC_DESCRIPTION_OFFER, (data: Types.SDP) => {
+            this.localConnection.setRemoteDescription(data.sdp)
+                .then(() => {
+                    console.log('Receiver remote SDP set.');
+                    this.localConnection.createAnswer()
+                        .then((answer: RTCSessionDescriptionInit) => {
+                            this.localConnection.setLocalDescription(answer);
+                            socket.emit(Constants.RTC_DESCRIPTION_ANSWER, {
+                                sdp: this.localConnection.localDescription
+                            });
+                        })
+                        .catch(this.handleCreateAnswerError);
+                })
+                .catch(this.handleSetDescriptionError);
+        });
+
+        socket.on(Constants.RTC_DESCRIPTION_ANSWER, (data: Types.SDP) => {
+            this.localConnection.setRemoteDescription(data.sdp)
+                .then(() => {
+                    console.log('Sender remote SDP set.');
+                })
+                .catch(this.handleSetDescriptionError);
+        });
+
+        socket.on(Constants.ICE_CANDIDATE, (data: RTCIceCandidate) => {
+            this.localConnection.addIceCandidate(data).catch(this.handleAddCandidateError);
+        });
+    } 
+    
+    /**
+     * Creates data channel and if initiator, creates offer and sends it over socket to room.
+     * If sendChannel and receiveChannel does not open up in TIMEOUT_MS, attempt to reconnect.
+     */
+    connectPeers = (channel: string, initiator: boolean): void => {
+        this.sendChannel = this.localConnection.createDataChannel(channel);
+        this.sendChannel.onopen = this.handleSendChannelStatusChange;
+        this.sendChannel.onclose = this.handleSendChannelStatusChange;
+
+        this.localConnection.ondatachannel = this.receiveChannelCallback;
+        this.localConnection.onicecandidate = this.handleOnICECandidate;
+
+        if (initiator) {
+            this.localConnection
+                .createOffer()
+                .then((offer: RTCSessionDescriptionInit) => this.localConnection.setLocalDescription(offer))
+                .then(() => {
+                    console.log('Sender SDP sent.');
+                    socket.emit(Constants.RTC_DESCRIPTION_OFFER, {
+                        sdp: this.localConnection.localDescription
+                    });
+                })
+                .catch(this.handleCreateDescriptionError);
+        }
+
+        setTimeout(() => {
```

Comment:
Is there a max number of retries on this? Won't it be an infinite loop if one of the users just leaves and we keep retrying?

Reply:
When a user leaves, `componentWillUnmount` is called on every user in the room since the `Messaging` or `FileTransfer` component would not be loaded since the room is not full anymore. But I guess this is assuming that there are only two people in the room. We should change the connect/disconnect logic to account for multiple users. Totally forgot that we were handling > 2 users.

Reply:
Ya maybe we should limit the number of retries to 3? Or any reasonable number?